### DISCUSSION TITLE
TONNE KM minimization

### DISCUSSION THREAD

**Discussion by surendrark29** [google/or-tools#2671] [Routing (and legacy CP) questions]
Hi,

I am working on a problem where I need to minimize the total tonne-km. 
As per my understanding OR tools take the distance matrix (or time matrix depending on objective) and finds routes that minimize the distance.

Can we track the total demand carried by each vehicle on each link? If yes, how can we do it?
If we can track, can we multiply it with link distance and use this data to minimize rather than just distance?

**Comment by lperron** [comment]
You need to use the Soft penalties on the cumul vars of the node.
Use a dimension that accumulate km without slack and with starts at 0.

At node k, it will contains the number of km done since the depot.
Add a soft upper bound with target 0 and value the tonnage.

The cost will then be distance travelled since depot * the tonnage.

**Reply by surendrark29** [reply]
Hi lperron,

Thank you for the reply. 
The following is the code I have used based on your reply.

```python
	def distance_callback(from_index, to_index):
		from_node = manager.IndexToNode(from_index)
		to_node = manager.IndexToNode(to_index)
		
		return data['distance_matrix'][from_node][to_node]

        dimension_name = 'Distance'

	routing.AddDimension(
		dist_callback_index,
		0,  # no slack
		20000000,  # vehicle maximum travel distance
		True,  # start cumul to zero
		dimension_name)

	distance_dimension = routing.GetDimensionOrDie(dimension_name)

	def demand_callback(from_index):
		from_node = manager.IndexToNode(from_index)
		return data['demands'][from_node]

	capacity = 'Capacity'

	demand_callback_index = routing.RegisterUnaryTransitCallback(
		demand_callback)

	routing.AddDimensionWithVehicleCapacity(
		demand_callback_index,
		0,  # null capacity slack
		data['vehicle_capacities'],  # vehicle maximum capacities
		True,  # start cumul to zero
		capacity)

	dim_dimension = routing.GetDimensionOrDie(capacity)

	for order in data['orders']:
		order_index = manager.NodeToIndex(order)
		distance_dimension.SetCumulVarSoftUpperBound(order_index, 0, dim_dimension.CumulVar(order_index))
```

When I tried running this I'm getting TypeError - I think this is due to the tonnage part - dim_dimension.CumulVar(order_index)

Can you please let me know where i'm going wrong.

**Comment by lperron** [comment]
https://google.github.io/or-tools/java/classcom_1_1google_1_1ortools_1_1constraintsolver_1_1RoutingDimension.html#a8b07631c1066331b2ecfef7754be3198


Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by surendrark29** [reply]
Thanks for the prompt reply.
Based on the link - distance_dimension.SetCumulVarSoftUpperBound(order_index, 0, dim_dimension.CumulVar(order_index)) is correct
I want to multiply all distances with tonnage (given by dim_dimension.CumulVar(order_index))

only thing is tonnage is a variable (based on the route) not a constant, which is leading to the issue

**Comment by lperron** [comment]
for one demand, the tonnage is constant. You need to add this soft upper
bound on the delivery node of the demand.
And do this for each demand.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by surendrark29** [reply]
even after using a constant it is still throwing the following error
TypeError: in method 'RoutingModel_AddDimension', argument 4 of type 'int64'

**Reply by Mizux** [reply]
does your constant is an integer ?

**Reply by surendrark29** [reply]
I passed the value after converting to integer using int(value).

I have even tried using the following - distance_dimension.SetCumulVarSoftUpperBound(order_index, 0, 100)
it still shows the same error
